src/SFA.DAS.EmployerFinance/Models/LevyDeclarationSaga.cs

```diff
@@ -0,0 +1,95 @@
+            ImportPayeSchemeLevyDeclarationsTasksCount = accountPayeSchemes.Select(aps => aps.EmployerReferenceNumber).Distinct().Count();
+            UpdateAccountTransactionBalancesTasksCount = accountPayeSchemes.Select(aps => aps.AccountId).Distinct().Count();
+            
+            Publish(() => new StartedProcessingLevyDeclarationsEvent(Id, PayrollPeriod, Created));
+        }
+
+        public LevyDeclarationSaga(DateTime payrollPeriod, AccountPayeScheme accountPayeScheme)
+        {
+            Type = LevyDeclarationSagaType.AdHoc;
+            PayrollPeriod = payrollPeriod;
+            Created = DateTime.UtcNow;
+            HighWaterMarkId = accountPayeScheme.Id;
+            ImportPayeSchemeLevyDeclarationsTasksCount = 1;
+            UpdateAccountTransactionBalancesTasksCount = 1;
+            
+            Publish(() => new StartedProcessingLevyDeclarationsAdHocEvent(Id, PayrollPeriod, HighWaterMarkId, Created));
+        }
+
+        private LevyDeclarationSaga()
+        {
+        }
+        
+        public void UpdateProgress(IReadOnlyCollection<LevyDeclarationSagaTask> tasks)
+        {
+            if (!IsComplete)
+            {
+                if (ImportPayeSchemeLevyDeclarationsTasksCompleteCount < ImportPayeSchemeLevyDeclarationsTasksCount)
+                {
+                    ImportPayeSchemeLevyDeclarationsTasksCompleteCount = tasks.Count(t => t.Type == LevyDeclarationSagaTaskType.ImportPayeSchemeLevyDeclarations);
+                    Updated = DateTime.UtcNow;
+
+                    if (ImportPayeSchemeLevyDeclarationsTasksCompleteCount == ImportPayeSchemeLevyDeclarationsTasksCount)
+                    {
+                        ImportPayeSchemeLevyDeclarationsTasksFinished = Updated;
+                        
+                        Publish(() => new UpdatedLevyDeclarationSagaProgressEvent(Id));
+                    }
+                }
+                else
+                {
+                    UpdateAccountTransactionBalancesTasksCompleteCount = tasks.Count(t => t.Type == LevyDeclarationSagaTaskType.UpdateAccountTransactionBalances);
+                    Updated = DateTime.UtcNow;
+
+                    if (UpdateAccountTransactionBalancesTasksCompleteCount == UpdateAccountTransactionBalancesTasksCount)
+                    {
+                        UpdateAccountTransactionBalancesTasksFinished = Updated;
+                        IsComplete = true;
+
+                        switch (Type)
+                        {
+                            case LevyDeclarationSagaType.All:
+                                Publish(() => new FinishedProcessingLevyDeclarationsEvent(Id, PayrollPeriod, Updated.Value));
+                                break;
+                            case LevyDeclarationSagaType.AdHoc:
+                                Publish(() => new FinishedProcessingLevyDeclarationsAdHocEvent(Id, PayrollPeriod, HighWaterMarkId, Updated.Value));
```

Comment:
This is correct, but the naming of the ad hoc id being processed as the `HighWaterMarkId` initially confused me. Could we rename `HighWaterMarkId` to better represent its two uses, or else have two properties, say, HighWaterMarkId and AdHocAccountPayeSchemeId or similar.

Reply:
Agreed, I've added `AccountLegalEntityHighWaterMarkId` & `AccountLegalEntityId` properties.